homeassistant/components/zwave_js/discovery_data_template.py

```diff
@@ -251,3 +251,102 @@ def values_to_watch(self, resolved_data: dict[str, Any]) -> Iterable[ZwaveValue]
+        Get the fan speed configuration for this device.
+
+        Values should indicate the highest allowed device setting for each
+        actual speed.
+        """
+        # pylint: disable=no-self-use
+        return None
+
+
+@dataclass
+class FixedFanSpeedDataTemplate(BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate):
+    """
+    Specifies a fixed set of fan speeds.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=FixedFanSpeedDataTemplate(
+              speeds=[32,65,99]
+          ),
+      ),
+
+    `speeds` indicates the maximum setting on the underlying fan controller
+    for each actual speed.
+    """
+
+    speeds: list[int] = field(default_factory=list)
+
+    def get_speed_config(self, resolved_data: dict[str, Any]) -> list[int] | None:
+        """Get the fan speed configuration for this device."""
+        return self.speeds
+
+
+@dataclass
+class ConfigurableFanSpeedDataTemplate(
+    BaseDiscoverySchemaDataTemplate, FanSpeedDataTemplate
+):
+    """
+    Gets fan speeds based on a configuration value.
+
+    Example:
+      ZWaveDiscoverySchema(
+          platform="fan",
+          hint="configured_fan_speed",
+          ...
+          data_template=ConfigurableFanSpeedDataTemplate(
+            configuration_option=ZwaveValueID(
+                5, CommandClass.CONFIGURATION, endpoint=0
+            ),
+            configuration_value_to_speeds={0: [32, 65, 99], 1: [24, 49, 74, 99]},
+          ),
+      ),
+
+    `configuration_option` is a reference to the setting that determines how
+    many speeds are supported.
+
+    `configuration_value_to_speeds` maps the values from `configuration_option`
+    to a list of speeds.  The specified speeds indicate the maximum setting on
+    the underlying switch for each actual speed.
+    """
+
+    configuration_option: ZwaveValueID | None = None
+    configuration_value_to_speeds: dict[int, list[int]] = field(default_factory=dict)
+
+    def resolve_data(self, value: ZwaveValue) -> dict[str, Any]:
+        """Resolve helper class data for a discovered value."""
+        if not self.configuration_option:
+            raise ValueError("Invalid discovery data template")
+
+        data: dict[str, Any] = {
+            "configuration_value": self._get_value_from_id(
+                value.node, self.configuration_option
+            ),
+        }
+        return data
+
+    def values_to_watch(self, resolved_data: dict[str, Any]) -> Iterable[ZwaveValue]:
+        """Return list of all ZwaveValues that should be watched."""
+        return [
+            resolved_data["configuration_value"],
+        ]
+
+    def get_speed_config(self, resolved_data: dict[str, Any]) -> list[int] | None:
+        """Get current speed configuration from resolved data."""
+        configuration_value: ZwaveValue | None = resolved_data["configuration_value"]
+        if configuration_value:
```

Comment:
When can the value be None? Isn't that just a developer mistake and we should blow up in that case?

Reply:
I haven't entirely figured out what's happening with this, but there's a brief moment during HA's startup where the device's config data doesn't seem to be available and this method returns `None`.

Which isn't to say that this isn't a developer mistake (possibly mine, even), but carefully handling that case lets us avoid having a bunch of errors in the logs during startup.

Reply:
I've _kinda_ figured out what was happening. I've got 4 of these devices, and just one of them was malfunctioning (not the one I've been testing with).  I'd previously configured it to be in 4-speed mode, but sometime afterwards it seems to have stopped reporting its config values (even in HA's device config UI).  They seemed to still be in full effect on the device though (i.e., on the device itself I could step through all 4 speeds). But once I went in and reset the value in the HA device config UI, the config value started being reported back as well.

We've got a few options for handling this sort of error:
a) Make the device unavailable until the config values are fixed (the current implementation)
b) Fall back to 1% steps until the config is fixed (probably requires merging ConfiguredSpeedRangeZwaveFan with the base ZwaveFan so we can dynamically switch between modes at runtime)
c) Fall back to one of the pre-configured speeds

In any of those cases, I think we'd want some kind of logging or notification that the device config is borked. Thoughts?

Reply:
@raman325 @kpine what do you think? I'm still surprised that all values are not present if the node is ready. Is that specific for configuration values?

Reply:
Was it the "value" that was missing, or the "value id"?

Reply:
I was able to get the 'Value' (a 'ConfigurationValue', via _get_value_from_id), but its 'value' property wasn't set.  This condition also survived across many restarts of HA until I re-set the value in HA's UI.

Reply:
So I power-cycled the device, and restarted zwave-js-server (in retrospect I should have done one at a time), and that seems to have completely fixed the configuration settings (-- the ones that I hadn't previously reset).

Reply:
Z-Wave JS doesn't refresh configuration parameters on startup, so wouldn't expect the power cycle or server restart to have done anything. Do you have any debug logs of the startup?

As to why there was a value ID, but not a value? I'm not sure, maybe during the original interview there were errors reading the values, so it was never populated? I checked my FC200 and it's fine.

Reply:
I've pushed the relevant zwave-js-server startup log to https://gist.github.com/mkowalchuk/1528d6d2d52045f4ebac2516ace6f21d startup log.  Node 061 is the FC200 that was misbehaving.  Nodes 039, 045, 066 are the other FC200s.  I don't see any smoking gun in there though.

Reply:
> I was able to get the 'Value' (a 'ConfigurationValue', via _get_value_from_id), but its 'value' property wasn't set. This condition also survived across many restarts of HA until I re-set the value in HA's UI.

Note that it's expected to not have the Value.value property return something other than None until later after interview but the Value should not be None when the node is ready, as far as I know.

We should then be able to remove the check here.

Reply:
Alright, I've added some harder failures (exceptions) in the data resolver, and removed the checks in the component.